README.md
TRIVECHAIN CORE 2.3.0.0
===============================
Trivechain is a public blockchain managed by Decentralized autonomous organization (DAO) with focus on games and entertainment with implementation of blockchain-based technology and DApps (decentralized applications)

Visit https://trivechain.com to learn more

Blockchain Specs
----------------
<table>
<tr><td>Genesis Block Time</td><td>1st March 2018</td></tr>
<tr><td>Decimals</td><td>8</td></tr>
<tr><td>Block Time</td><td>60 Seconds</td></tr>
<tr><td>Difficulty Retargeting</td><td>Every Block</td></tr>
<tr><td>Initial Block Reward (TRVC 2.0 Era)</td><td>25 TRVC</td></tr>
<tr><td>Block Reward Adjustment (-25%)</td><td>Every 525,600 Blocks / 1 year</td></tr>
<tr><td>Proof of Work (PoW) Algo</td><td>X16Rv2</td></tr>
<tr><td>Masternode (PoSe) Collateral</td><td>10,000 TRVC</td></tr>
<tr><td>TriveGovernance Proposal Fee</td><td>50 TRVC</td></tr>
<tr><td>Max Supply (Best Estimation)</td><td>80,582,104 TRVC</td></tr>
</table>

Useful Links
----------------
Official Website
- Trivechain: https://trivechain.com/
- TriveAcademy: https://trive.academy/

Explorer
- Blockchain Explorer 1 - [https://explorer.trivechain.com](https://explorer.trivechain.com)
- Blockchain Explorer 2 - [https://insight.trivechain.com](https://insight.trivechain.com)

Wallet
- Trivechain’s wallet (TRVC App): [https://trvc.app](https://trvc.app)
- Trivechain’s copay wallet: [https://wallet.trivechain.com](https://wallet.trivechain.com)

Mining Pool (Pow)
- Official Pool: [https://pool.trivechain.com](https://pool.trivechain.com)

PoW Mining
- Nvidia CUDA Graphic Card: [https://github.com/trexminer/T-Rex/releases](https://github.com/trexminer/T-Rex/releases)
- CPU: [https://github.com/tpruvot/cpuminer-multi](https://github.com/tpruvot/cpuminer-multi)
- MiningRigRentals: [https://www.miningrigrentals.com](https://www.miningrigrentals.com/?ref=2609423)

Masternode Setup (PoSe)
- Masternode Registration Guide: https://github.com/trivechain/trivechain-core/blob/master/docs/masternode-registration.md
- Masternode Renting Platform 1: https://citihash.com

Social Media
- Discord: https://discord.gg/jhbaf5E
- Facebook: https://www.facebook.com/trivechain/
- Twitter: https://twitter.com/trivechain_trvc
- Telegram: https://t.me/trivechain
- Wechat: TRVC20
- Medium: https://medium.com/trivechain
- Youtube: https://www.youtube.com/channel/UCIUtsQZ8KCWWD9M3CpEAR-w
- Reddit: https://www.reddit.com/user/TRVC-2

Exchanger
- BitRabbit: https://bitrabbit.com/markets/trvc_btc
- PARS Exchange: https://parsxc.com/trade/index/market/trvc_btc/

Testnet ( v2 Genesis started on 22 Oct 2019 )
- Pool: https://pool.trvc.dev
- Blockchain Explorer 1: https://insight.trvc.dev
- Blockchain Explorer 2: https://explorer.trvc.dev/

Block Rewards and Circulating Supply
----------------
<table>
<th>Year</th><th>Block Height</th><th>Miners (42%)</th><th>Masternodes (28%)</th><th>TriveGovernance (30%)</th><th>Circulating Supply</th><tr><td>1</td><td>1 - 226655</td><td>-</td><td>-</td><td>-</td><td>29,986,584.92019400</td></tr>
<tr><td>2</td><td>226655 - 448299</td><td>10.50000000</td><td>7.00000000</td><td>0.00000000</td><td>33,865,354.92019400</td></tr>
</table>
<table>
<th>Year</th><th>Block Height</th><th>Miners (35%)</th><th>Masternodes (35%)</th><th>TriveGovernance (30%)</th><th>Circulating Supply</th>
<tr><td>2</td><td>448300 - 525599</td><td>8.75000000</td><td>8.75000000</td><td>0.00000000</td><td>35,495,830.92019440</td></tr>
<tr><td>2</td><td>525600 - 752255</td><td>8.75000000</td><td>8.75000000</td><td>7.50000000</td><td>41,162,180.92019440</td></tr>
<tr><td>3</td><td>752255 - 1277854</td><td>6.56250000</td><td>6.56250000</td><td>5.62500000</td><td>51,017,162.17019440</td></tr>
<tr><td>4</td><td>1277855 - 1803454</td><td>4.92187499</td><td>4.92187499</td><td>4.21875000</td><td>58,408,398.10769440</td></tr>
<tr><td>5</td><td>1803455 - 2329054</td><td>3.69140625</td><td>3.69140625</td><td>3.16406250</td><td>63,951,825.06081940</td></tr>
<tr><td>6</td><td>2329055 - 2854654</td><td>2.76855468</td><td>2.76855468</td><td>2.37304687</td><td>68,109,395.27566315</td></tr>
<tr><td>7</td><td>2854655 - 3380254</td><td>2.07641601</td><td>2.07641601</td><td>1.77978515</td><td>71,227,572.93285397</td></tr>
<tr><td>8</td><td>3380255 - 3905854</td><td>1.55731200</td><td>1.55731200</td><td>1.33483886</td><td>73,566,206.17311908</td></tr>
<tr><td>9</td><td>3905855 - 4431454</td><td>1.16798400</td><td>1.16798400</td><td>1.00112914</td><td>75,320,181.10331793</td></tr>
<tr><td>10</td><td>4431455 - 4957054</td><td>0.87598800</td><td>0.87598800</td><td>0.75084686</td><td>76,635,662.30096705</td></tr>
<tr><td>11</td><td>4957055 - 5482654</td><td>0.65699100</td><td>0.65699100</td><td>0.56313514</td><td>77,622,273.19788991</td></tr>
<tr><td>12</td><td>5482655 - 6008254</td><td>0.49274325</td><td>0.49274325</td><td>0.42235135</td><td>78,362,231.36926804</td></tr>
<tr><td>13</td><td>6008255 - 6533854</td><td>0.36955743</td><td>0.36955743</td><td>0.31676351</td><td>78,917,199.99517366</td></tr>
<tr><td>14</td><td>6533855 - 7059454</td><td>0.27716807</td><td>0.27716807</td><td>0.23757263</td><td>79,333,426.46328887</td></tr>
<tr><td>15</td><td>7059455 - 7585054</td><td>0.20787605</td><td>0.20787605</td><td>0.17817947</td><td>79,645,596.31306128</td></tr>
<tr><td>16</td><td>7585055 - 8110654</td><td>0.15590704</td><td>0.15590704</td><td>0.13363460</td><td>79,879,723.69907659</td></tr>
<tr><td>17</td><td>8110655 - 8636254</td><td>0.11693027</td><td>0.11693027</td><td>0.10022595</td><td>80,055,319.23464608</td></tr>
<tr><td>18</td><td>8636255 - 9161854</td><td>0.08769770</td><td>0.08769770</td><td>0.07516946</td><td>80,187,015.88500920</td></tr>
<tr><td>19</td><td>9161855 - 9687454</td><td>0.06577328</td><td>0.06577328</td><td>0.05637709</td><td>80,285,788.37278154</td></tr>
<tr><td>20</td><td>9687455 - 10213054</td><td>0.04932995</td><td>0.04932995</td><td>0.04228282</td><td>80,359,867.73598281</td></tr>
<tr><td>21</td><td>10213055 - 10738654</td><td>0.03699746</td><td>0.03699746</td><td>0.03171211</td><td>80,415,427.25575575</td></tr>
<tr><td>22</td><td>10738655 - 11264254</td><td>0.02774809</td><td>0.02774809</td><td>0.02378408</td><td>80,457,096.89164346</td></tr>
<tr><td>23</td><td>11264255 - 11789854</td><td>0.02081107</td><td>0.02081107</td><td>0.01783806</td><td>80,488,349.11855926</td></tr>
<tr><td>24</td><td>11789855 - 12315454</td><td>0.01560830</td><td>0.01560830</td><td>0.01337854</td><td>80,511,788.28480411</td></tr>
<tr><td>25</td><td>12315455 - 12841054</td><td>0.01170622</td><td>0.01170622</td><td>0.01003390</td><td>80,529,367.65817375</td></tr>
<tr><td>26</td><td>12841055 - 13366654</td><td>0.00877966</td><td>0.00877966</td><td>0.00752543</td><td>80,542,552.18820098</td></tr>
<tr><td>27</td><td>13366655 - 13892254</td><td>0.00658474</td><td>0.00658474</td><td>0.00564407</td><td>80,552,440.58177941</td></tr>
<tr><td>28</td><td>13892255 - 14417854</td><td>0.00493855</td><td>0.00493855</td><td>0.00423305</td><td>80,559,856.87302125</td></tr>
<tr><td>29</td><td>14417855 - 14943454</td><td>0.00370391</td><td>0.00370391</td><td>0.00317478</td><td>80,565,419.08751062</td></tr>
<tr><td>30</td><td>14943455 - 15469054</td><td>0.00277793</td><td>0.00277793</td><td>0.00238108</td><td>80,569,590.74574967</td></tr>
<tr><td>31</td><td>15469055 - 15994654</td><td>0.00208345</td><td>0.00208345</td><td>0.00178581</td><td>80,572,719.48942894</td></tr>
<tr><td>32</td><td>15994655 - 16520254</td><td>0.00156258</td><td>0.00156258</td><td>0.00133936</td><td>80,575,066.04718840</td></tr>
<tr><td>33</td><td>16520255 - 17045854</td><td>0.00117193</td><td>0.00117193</td><td>0.00100452</td><td>80,576,825.96288000</td></tr>
<tr><td>34</td><td>17045855 - 17571454</td><td>0.00087895</td><td>0.00087895</td><td>0.00075339</td><td>80,578,145.89964870</td></tr>
<tr><td>35</td><td>17571455 - 18097054</td><td>0.00065921</td><td>0.00065921</td><td>0.00056504</td><td>80,579,135.84959723</td></tr>
<tr><td>36</td><td>18097055 - 18622654</td><td>0.00049441</td><td>0.00049441</td><td>0.00042378</td><td>80,579,878.31074463</td></tr>
<tr><td>37</td><td>18622655 - 19148254</td><td>0.00037080</td><td>0.00037080</td><td>0.00031783</td><td>80,580,435.15660518</td></tr>
<tr><td>38</td><td>19148255 - 19673854</td><td>0.00027810</td><td>0.00027810</td><td>0.00023837</td><td>80,580,852.78705861</td></tr>
<tr><td>39</td><td>19673855 - 20199454</td><td>0.00020857</td><td>0.00020857</td><td>0.00017877</td><td>80,581,166.00727066</td></tr>
<tr><td>40</td><td>20199455 - 20725054</td><td>0.00015642</td><td>0.00015642</td><td>0.00013408</td><td>80,581,400.91848773</td></tr>
<tr><td>41</td><td>20725055 - 21250654</td><td>0.00011732</td><td>0.00011732</td><td>0.00010056</td><td>80,581,577.09927253</td></tr>
<tr><td>42</td><td>21250655 - 21776254</td><td>0.00008799</td><td>0.00008799</td><td>0.00007542</td><td>80,581,709.23486114</td></tr>
<tr><td>43</td><td>21776255 - 22301854</td><td>0.00006599</td><td>0.00006599</td><td>0.00005656</td><td>80,581,808.33655258</td></tr>
<tr><td>44</td><td>22301855 - 22827454</td><td>0.00004949</td><td>0.00004949</td><td>0.00004242</td><td>80,581,882.66150717</td></tr>
<tr><td>45</td><td>22827455 - 23353054</td><td>0.00003711</td><td>0.00003711</td><td>0.00003181</td><td>80,581,938.40128112</td></tr>
<tr><td>46</td><td>23353055 - 23878654</td><td>0.00002783</td><td>0.00002783</td><td>0.00002385</td><td>80,581,980.20216960</td></tr>
<tr><td>47</td><td>23878655 - 24404254</td><td>0.00002087</td><td>0.00002087</td><td>0.00001789</td><td>80,582,011.54889394</td></tr>
<tr><td>48</td><td>24404255 - 24929854</td><td>0.00001565</td><td>0.00001565</td><td>0.00001341</td><td>80,582,035.05893722</td></tr>
<tr><td>49</td><td>24929855 - 25455454</td><td>0.00001173</td><td>0.00001173</td><td>0.00001006</td><td>80,582,052.68752769</td></tr>
<tr><td>50</td><td>25455455 - 25981054</td><td>0.00000880</td><td>0.00000880</td><td>0.00000754</td><td>80,582,065.90634254</td></tr>
<tr><td>51</td><td>25981055 - 26506654</td><td>0.00000660</td><td>0.00000660</td><td>0.00000565</td><td>80,582,075.81913967</td></tr>
<tr><td>52</td><td>26506655 - 27032254</td><td>0.00000494</td><td>0.00000494</td><td>0.00000424</td><td>80,582,083.25110953</td></tr>
<tr><td>53</td><td>27032255 - 27557854</td><td>0.00000371</td><td>0.00000371</td><td>0.00000318</td><td>80,582,088.82245892</td></tr>
<tr><td>54</td><td>27557855 - 28083454</td><td>0.00000278</td><td>0.00000278</td><td>0.00000238</td><td>80,582,093.00097097</td></tr>
<tr><td>55</td><td>28083455 - 28609054</td><td>0.00000208</td><td>0.00000208</td><td>0.00000178</td><td>80,582,096.13354102</td></tr>
<tr><td>56</td><td>28609055 - 29134654</td><td>0.00000156</td><td>0.00000156</td><td>0.00000134</td><td>80,582,098.48296855</td></tr>
<tr><td>57</td><td>29134655 - 29660254</td><td>0.00000117</td><td>0.00000117</td><td>0.00000100</td><td>80,582,100.24372520</td></tr>
<tr><td>58</td><td>29660255 - 30185854</td><td>0.00000087</td><td>0.00000087</td><td>0.00000075</td><td>80,582,101.56297868</td></tr>
<tr><td>59</td><td>30185855 - 30711454</td><td>0.00000065</td><td>0.00000065</td><td>0.00000056</td><td>80,582,102.55110481</td></tr>
<tr><td>60</td><td>30711455 - 31237054</td><td>0.00000049</td><td>0.00000049</td><td>0.00000042</td><td>80,582,103.29219940</td></tr>
<tr><td>61</td><td>31237055 - 31762654</td><td>0.00000036</td><td>0.00000036</td><td>0.00000031</td><td>80,582,103.84407835</td></tr>
<tr><td>62</td><td>31762655 - 32288254</td><td>0.00000027</td><td>0.00000027</td><td>0.00000023</td><td>80,582,104.25404558</td></tr>
</table>

License
-------
Trivechain is released under the terms of the MIT license. See [COPYING](COPYING) for more
information or see https://opensource.org/licenses/MIT.

Development Process
-------------------
The `master` branch is meant to be stable. Development is normally done in separate branches.
[Tags](https://github.com/trivechain/trivechain-core/tags) are created to indicate new official,
stable release versions of Trivechain.

The contribution workflow is described in [CONTRIBUTING.md](CONTRIBUTING.md).


docs/release-notes.md
Trivechain Core version 0.15
======================

Release is now available from:

  <https://www.trivechain.com/downloads/#wallets>

This is a new major version release, bringing new features, various bugfixes and other improvements.

Please report bugs using the issue tracker at github:

  <https://github.com/trivechain/trivechain/issues>


Upgrading and downgrading
=========================

How to Upgrade
--------------

If you are running an older version, shut it down. Wait until it has completely
shut down (which might take a few minutes for older versions), then run the
installer (on Windows) or just copy over /Applications/Trivechain-Qt (on Mac) or
trivechaind/trivechain-qt (on Linux). If you upgrade after DIP0003 activation and you were
using version < 0.13 you will have to reindex (start with -reindex-chainstate
or -reindex) to make sure your wallet has all the new data synced. Upgrading from
version 0.13 should not require any additional actions.

When upgrading from a version prior to 0.14.0.3, the
first startup of Trivechain Core will run a migration process which can take a few minutes
to finish. After the migration, a downgrade to an older version is only possible with
a reindex (or reindex-chainstate).

Downgrade warning
-----------------

### Downgrade to a version < 0.14.0.3

Downgrading to a version smaller than 0.14.0.3 is not supported anymore due to changes
in the "evodb" database format. If you need to use an older version, you have to perform
a reindex or re-sync the whole chain.

Notable changes
===============

Removal of the p2p alert system
-------------------------------
The p2p alert system was designed to send messages to all nodes supporting it by someone who holds
so called alert keys to notify such nodes in case of severe network issues. This version removes
the `alert` p2p message and `--alert` option. Internal alerts, partition detection warnings and the
`--alertnotify` option features remain.

Removal of the legacy DirectSend system
----------------------------------------
Version 0.14 introduced the new LLMQ-based DirectSend system which is designed to be much more scalable
than the legacy one without sacrificing security. The new system also allows all transactions to be treated as
DirectSend transactions. The legacy system was disabled together with the successful deployment of ChainLocks,
but we had to keep supporting the legacy system for a while to ensure a smooth transition period. This version finally drops
the legacy system completely.

Read more about ChainLocks: https://github.com/trivechain/dips/blob/master/dip-0008.md
Read more about LLMQ-based DirectSend: https://github.com/trivechain/dips/blob/master/dip-0010.md

Sporks
------
The security level of ChainLocks and LLMQ-based DirectSend made sporks `SPORK_5_DIRECTSEND_MAX_VALUE` and
`SPORK_12_RECONSIDER_BLOCKS` obsolete, so they are removed now. Sporks `SPORK_15_DETERMINISTIC_MNS_ENABLED`,
`SPORK_16_DIRECTSEND_AUTOLOCKS` and `SPORK_20_DIRECTSEND_LLMQ_BASED` have no code logic behind them anymore
because they were used as part of the DIP0003, DIP0008 and DIP0010 activation process which is finished now.
They are still kept and relayed only to ensure smooth operation of v0.14 clients and will be removed in some
future version.

Mempool sync improvements
-------------------------
Nodes joining the network will now try to sync their mempool from other v0.15+ peers via the `mempool` p2p message.
This behaviour can be disabled via the new `--syncmempool` option. Nodes serving such requests will now also push
`inv` p2p messages for InstandSend locks which are held for transactions in their mempool. These two changes
should help new nodes to quickly catchup on start and detect any potential double-spend as soon as possible.
This should also help wallets to slightly improve UX by showing the correct status of unconfirmed transactions
locked via InstandSend, if they were sent while the receiving wallet was offline. Note that bloom-filters
still apply to such `inv` messages, just like they do for transactions and locks that are relayed on a
regular basis.

PrivateSend improvements
------------------------
This version decouples the so called "Lite Mode" and client-side PrivateSend mixing, which allows client-side mixing
on pruned nodes running with `--litemode` option. Such nodes will have to also specify the newly redefined
`--enableprivatesend` option. Non-prunned nodes do not have to do this but they can use `--enableprivatesend`
option to disable mixing completely instead. Please note that specifying this option does not start mixing
automatically anymore (which was the case in previous versions). To automatically start mixing, use the new
`--privatesendautostart` option in addition to `--enableprivatesend`. Additionally, PrivateSend can always be
controlled with the `privatesend` RPC.

Thanks to LLMQ-based DirectSend and its ability to lock chains of unconfirmed transactions (and not only a single
one like in the legacy system), PrivateSend mixing speed has improved significantly. In such an environment
Liquidity Provider Mode, which was introduced a long time ago to support mixing volume, is no longer needed and
is removed now. As such the `--liquidityprovider` option is not available anymore.

Some other improvements were also introduced to speed up mixing, e.g. by joining more queues or dropping potential
malicious mixing participants faster by checking some rules earlier etc. Lots of related code was refactored to
further improve its readability, which should make it easier for someone to re-implement PrivateSend
correctly in other wallets if there is a desire to do so.

Wallet changes
--------------
Wallet internals were optimized to significantly improve performance which should be especially notable for huge
wallets with tens of thousands of transactions or more. The GUI for such wallets should be much more responsive too now.


src/util.h
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2015 The Bitcoin Core developers
// Copyright (c) 2014-2019 The Trivechain Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

/**
 * Server/client environment: argument handling, config file parsing,
 * logging, thread wrappers, startup time
 */
#ifndef BITCOIN_UTIL_H
#define BITCOIN_UTIL_H

#if defined(HAVE_CONFIG_H)
#include "config/trivechain-config.h"
#endif

#include "compat.h"
#include "fs.h"
#include "sync.h"
#include "tinyformat.h"
#include "utiltime.h"
#include "amount.h"

#include <atomic>
#include <exception>
#include <map>
#include <stdint.h>
#include <string>
#include <unordered_map>
#include <vector>

#include <boost/signals2/signal.hpp>

// Debugging macros

// Uncomment the following line to enable debugging messages
// or enable on a per file basis prior to inclusion of util.h
//#define ENABLE_TRIVECHAIN_DEBUG
#ifdef ENABLE_TRIVECHAIN_DEBUG
#define DBG( x ) x
#else
#define DBG( x )
#endif

//Trivechain only features

extern bool fMasternodeMode;
extern bool fLiteMode;
extern int nWalletBackups;

// Application startup time (used for uptime calculation)
int64_t GetStartupTime();

static const bool DEFAULT_LOGTIMEMICROS  = false;
static const bool DEFAULT_LOGIPS         = false;
static const bool DEFAULT_LOGTIMESTAMPS  = true;
static const bool DEFAULT_LOGTHREADNAMES = false;

/** Signals for translation. */
class CTranslationInterface
{
public:
    /** Translate a message to the native language of the user. */
    boost::signals2::signal<std::string (const char* psz)> Translate;
};

extern bool fPrintToConsole;
extern bool fPrintToDebugLog;

extern bool fLogTimestamps;
extern bool fLogTimeMicros;
extern bool fLogThreadNames;
extern bool fLogIPs;
extern std::atomic<bool> fReopenDebugLog;
extern CTranslationInterface translationInterface;

extern const char * const BITCOIN_CONF_FILENAME;
extern const char * const BITCOIN_PID_FILENAME;

extern std::atomic<uint64_t> logCategories;

/**
 * Translation function: Call Translate signal on UI interface, which returns a boost::optional result.
 * If no translation slot is registered, nothing is returned, and simply return the input.
 */
inline std::string _(const char* psz)
{
    boost::optional<std::string> rv = translationInterface.Translate(psz);
    return rv ? (*rv) : psz;
}

void SetupEnvironment();
bool SetupNetworking();

struct CLogCategoryActive
{
    std::string category;
    bool active;
};

namespace BCLog {
    enum LogFlags : uint64_t {
        NONE        = 0,
        NET         = (1 <<  0),
        TOR         = (1 <<  1),
        MEMPOOL     = (1 <<  2),
        HTTP        = (1 <<  3),
        BENCHMARK   = (1 <<  4),
        ZMQ         = (1 <<  5),
        DB          = (1 <<  6),
        RPC         = (1 <<  7),
        ESTIMATEFEE = (1 <<  8),
        ADDRMAN     = (1 <<  9),
        SELECTCOINS = (1 << 10),
        REINDEX     = (1 << 11),
        CMPCTBLOCK  = (1 << 12),
        RANDOM      = (1 << 13),
        PRUNE       = (1 << 14),
        PROXY       = (1 << 15),
        MEMPOOLREJ  = (1 << 16),
        LIBEVENT    = (1 << 17),
        COINDB      = (1 << 18),
        QT          = (1 << 19),
        LEVELDB     = (1 << 20),

        //Start Trivechain
        CHAINLOCKS  = ((uint64_t)1 << 32),
        GOBJECT     = ((uint64_t)1 << 33),
        DIRECTSEND = ((uint64_t)1 << 34),
        KEEPASS     = ((uint64_t)1 << 35),
        LLMQ        = ((uint64_t)1 << 36),
        LLMQ_DKG    = ((uint64_t)1 << 37),
        LLMQ_SIGS   = ((uint64_t)1 << 38),
        MNPAYMENTS  = ((uint64_t)1 << 39),
        MNSYNC      = ((uint64_t)1 << 40),
        PRIVATESEND = ((uint64_t)1 << 41),
        SPORK       = ((uint64_t)1 << 42),
        //End Trivechain

        ALL         = ~(uint64_t)0,
    };
}
static inline bool LogAcceptCategory(uint64_t category)
{
    return (logCategories.load(std::memory_order_relaxed) & category) != 0;
}

/** Returns a string with the log categories. */
std::string ListLogCategories();

/** Returns a string with the list of active log categories */
std::string ListActiveLogCategoriesString();

/** Returns a vector of the active log categories. */
std::vector<CLogCategoryActive> ListActiveLogCategories();

/** Return true if str parses as a log category and set the flags in f */
bool GetLogCategory(uint64_t *f, const std::string *str);

/** Send a string to the log output */
int LogPrintStr(const std::string &str);

/** Formats a string without throwing exceptions. Instead, it'll return an error string instead of formatted string. */
template<typename... Args>
std::string SafeStringFormat(const std::string& fmt, const Args&... args)
{
    try {
        return tinyformat::format(fmt, args...);
    } catch (std::runtime_error& fmterr) {
        std::string message = tinyformat::format("\n****TINYFORMAT ERROR****\n    err=\"%s\"\n    fmt=\"%s\"\n", fmterr.what(), fmt);
        fprintf(stderr, "%s", message.c_str());
        return message;
    }
}

/** Get format string from VA_ARGS for error reporting */
template<typename... Args> std::string FormatStringFromLogArgs(const char *fmt, const Args&... args) { return fmt; }

static inline void MarkUsed() {}
template<typename T, typename... Args> static inline void MarkUsed(const T& t, const Args&... args)
{
    (void)t;
    MarkUsed(args...);
}

#ifdef USE_COVERAGE
#define LogPrintf(...) do { MarkUsed(__VA_ARGS__); } while(0)
#define LogPrint(category, ...) do { MarkUsed(__VA_ARGS__); } while(0)
#else
#define LogPrintf(...) do { \
    std::string _log_msg_; /* Unlikely name to avoid shadowing variables */ \
    try { \
        _log_msg_ = tfm::format(__VA_ARGS__); \
    } catch (tinyformat::format_error &e) { \
        /* Original format string will have newline so don't add one here */ \
        _log_msg_ = "Error \"" + std::string(e.what()) + "\" while formatting log message: " + FormatStringFromLogArgs(__VA_ARGS__); \
    } \
    LogPrintStr(_log_msg_); \
} while(0)

#define LogPrint(category, ...) do { \
    if (LogAcceptCategory((category))) { \
        LogPrintf(__VA_ARGS__); \
    } \
} while(0)
#endif

template<typename... Args>
bool error(const char* fmt, const Args&... args)
{
    LogPrintStr("ERROR: " + SafeStringFormat(fmt, args...) + "\n");
    return false;
}

void PrintExceptionContinue(const std::exception_ptr pex, const char* pszThread);
void FileCommit(FILE *file);
bool TruncateFile(FILE *file, unsigned int length);
int RaiseFileDescriptorLimit(int nMinFD);
void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length);
bool RenameOver(fs::path src, fs::path dest);
bool TryCreateDirectories(const fs::path& p);
fs::path GetDefaultDataDir();
const fs::path &GetDataDir(bool fNetSpecific = true);
fs::path GetBackupsDir();
void ClearDatadirCache();
fs::path GetConfigFile(const std::string& confPath);
#ifndef WIN32
fs::path GetPidFile();
void CreatePidFile(const fs::path &path, pid_t pid);
#endif
#ifdef WIN32
fs::path GetSpecialFolderPath(int nFolder, bool fCreate = true);
#endif
void OpenDebugLog();
void ShrinkDebugFile();
void runCommand(const std::string& strCommand);

inline bool IsSwitchChar(char c)
{
#ifdef WIN32
    return c == '-' || c == '/';
#else
    return c == '-';
#endif
}

class ArgsManager
{
protected:
    CCriticalSection cs_args;
    std::map<std::string, std::string> mapArgs;
    std::map<std::string, std::vector<std::string> > mapMultiArgs;
public:
    void ParseParameters(int argc, const char*const argv[]);
    void ReadConfigFile(const std::string& confPath);
    std::vector<std::string> GetArgs(const std::string& strArg);

    /**
     * Return true if the given argument has been manually set
     *
     * @param strArg Argument to get (e.g. "-foo")
     * @return true if the argument has been set
     */
    bool IsArgSet(const std::string& strArg);

    /**
     * Return string argument or default value
     *
     * @param strArg Argument to get (e.g. "-foo")
     * @param strDefault (e.g. "1")
     * @return command-line argument or default value
     */
    std::string GetArg(const std::string& strArg, const std::string& strDefault);

    /**
     * Return integer argument or default value
     *
     * @param strArg Argument to get (e.g. "-foo")
     * @param nDefault (e.g. 1)
     * @return command-line argument (0 if invalid number) or default value
     */
    int64_t GetArg(const std::string& strArg, int64_t nDefault);

    /**
     * Return boolean argument or default value
     *
     * @param strArg Argument to get (e.g. "-foo")
     * @param fDefault (true or false)
     * @return command-line argument or default value
     */
    bool GetBoolArg(const std::string& strArg, bool fDefault);

    /**
     * Set an argument if it doesn't already have a value
     *
     * @param strArg Argument to set (e.g. "-foo")
     * @param strValue Value (e.g. "1")
     * @return true if argument gets set, false if it already had a value
     */
    bool SoftSetArg(const std::string& strArg, const std::string& strValue);

    /**
     * Set a boolean argument if it doesn't already have a value
     *
     * @param strArg Argument to set (e.g. "-foo")
     * @param fValue Value (e.g. false)
     * @return true if argument gets set, false if it already had a value
     */
    bool SoftSetBoolArg(const std::string& strArg, bool fValue);

    // Forces an arg setting. Called by SoftSetArg() if the arg hasn't already
    // been set. Also called directly in testing.
    void ForceSetArg(const std::string& strArg, const std::string& strValue);
    void ForceSetMultiArgs(const std::string& strArg, const std::vector<std::string>& values);
    void ForceRemoveArg(const std::string& strArg);
};

extern ArgsManager gArgs;

/**
 * Format a string to be used as group of options in help messages
 *
 * @param message Group name (e.g. "RPC server options:")
 * @return the formatted string
 */
std::string HelpMessageGroup(const std::string& message);

/**
 * Format a string to be used as option description in help messages
 *
 * @param option Option message (e.g. "-rpcuser=<user>")
 * @param message Option description (e.g. "Username for JSON-RPC connections")
 * @return the formatted string
 */
std::string HelpMessageOpt(const std::string& option, const std::string& message);

/**
 * Return the number of physical cores available on the current system.
 * @note This does not count virtual cores, such as those provided by HyperThreading
 * when boost is newer than 1.56.
 */
int GetNumCores();

void RenameThread(const char* name);
std::string GetThreadName();

namespace ctpl {
    class thread_pool;
}
void RenameThreadPool(ctpl::thread_pool& tp, const char* baseName);

/**
 * .. and a wrapper that just calls func once
 */
template <typename Callable> void TraceThread(const char* name,  Callable func)
{
    std::string s = strprintf("trivechain-%s", name);
    RenameThread(s.c_str());
    try
    {
        LogPrintf("%s thread start\n", name);
        func();
        LogPrintf("%s thread exit\n", name);
    }
    catch (const boost::thread_interrupted&)
    {
        LogPrintf("%s thread interrupt\n", name);
        throw;
    }
    catch (...) {
        PrintExceptionContinue(std::current_exception(), name);
        throw;
    }
}

std::string CopyrightHolders(const std::string& strPrefix, unsigned int nStartYear, unsigned int nEndYear);

/**
 * @brief Converts version strings to 4-byte unsigned integer
 * @param strVersion version in "x.x.x" format (decimal digits only)
 * @return 4-byte unsigned integer, most significant byte is always 0
 * Throws std::bad_cast if format doesn\t match.
 */
uint32_t StringVersionToInt(const std::string& strVersion);


/**
 * @brief Converts version as 4-byte unsigned integer to string
 * @param nVersion 4-byte unsigned integer, most significant byte is always 0
 * @return version string in "x.x.x" format (last 3 bytes as version parts)
 * Throws std::bad_cast if format doesn\t match.
 */
std::string IntVersionToString(uint32_t nVersion);


/**
 * @brief Copy of the IntVersionToString, that returns "Invalid version" string
 * instead of throwing std::bad_cast
 * @param nVersion 4-byte unsigned integer, most significant byte is always 0
 * @return version string in "x.x.x" format (last 3 bytes as version parts)
 * or "Invalid version" if can't cast the given value
 */
std::string SafeIntVersionToString(uint32_t nVersion);


#endif // BITCOIN_UTIL_H
